### DISCUSSION TITLE
how to run pytest.main() with thread

### DISCUSSION THREAD

**Discussion by zhengjiaoyang** [pytest-dev/pytest#10271] [Q&A] [answered]
```
def run_tasks():
    tasks = get_tasks(task_id) #task_list
    for task in tasks:
        t = threading.Thread(target=exec_task, args=(task,), name=task['id'])
        t.start()

def exec_task(task):
    pytest.main('-s', '-v', task['path'])
```
I want to run pytest main with thread in my services, but when len(tasks) > 1, the result of tasks always the task[0]. If I want run different project in parallel, what shall I do?

**Comment by RonnyPfannschmidt** [comment]
The outlined use case is absolutely unsupported, and won't be supported

Please outline the target use case for figuring a supported way

**Reply by zhengjiaoyang** [reply]
> The outlined use case is absolutely unsupported, and won't be supported
> 
> Please outline the target use case for figuring a supported way

there is different subproject
.
├── testproject1
│   └── test_case1.py
├── testproject2
│   └── test_case2.py
└── testproject3
    └── test_case3.py

and I want to to execute different project  through the API request.
Projects are independent of each other, when project1 is running,  the operation of project2 will not be affected

**Reply by RonnyPfannschmidt** [reply]
thats basically a folder structure, not a outlined usecase someone other than you can understand, please include more details about what this does

**Reply by zhengjiaoyang** [reply]
> 

I want to run multiple projects in parallel in one service.

These projects belong to different business lines, have different data sources and middleware, but each project needs to output an independent test report

In short, I just want to know if it is possible to run multiple pytest processes in one service at the same time. It has nothing to do with the details of the use case

**Reply by nicoddemus** [reply]
The usual solution to that is to run each "project" in a separate process, usually controlled by different CI jobs.

If you need to control that via a script for some reason, you should instead execute each `pytest.main` in its own subprocess, as this is safer. 

You can accomplish this by running via `subprocess.run(["pytest", ...])` or via `multiprocessing.Pool`, although I would recommend the former because I think this will carry less surprises.

**Reply by zhengjiaoyang** [reply]
> The usual solution to that is to run each "project" in a separate process, usually controlled by different CI jobs.
> 
> If you need to control that via a script for some reason, you should instead execute each `pytest.main` in its own subprocess, as this is safer.
> 
> You can accomplish this by running via `subprocess.run(["pytest", ...])` or via `multiprocessing.Pool`, although I would recommend the former because I think this will carry less surprises.

get， thanks

**Reply by nicoddemus** [reply]
(posted as a separate answer and marked as answered to help others finding the solution)

**Comment by nicoddemus** [comment, marked as answer]
The usual solution to that is to run each "project" in a separate process, usually controlled by different CI jobs.

If you need to control that via a script for some reason, you should instead execute each `pytest.main` in its own subprocess, as this is safer. 

You can accomplish this by running via `subprocess.run(["pytest", ...])` or via `multiprocessing.Pool`, although I would recommend the former because I think this will carry less surprises.
